File: client/test/game/host_sim_test.dart
// The host-side rules: night schedule, zombie attacks, sword and rifle,
// death → crate → respawn, poison, regen, drops from digging, placing.
import 'package:block_survival/entities/zombies.dart';
import 'package:block_survival/game/avatar.dart';
import 'package:block_survival/game/day_night.dart';
import 'package:block_survival/game/host_sim.dart';
import 'package:block_survival/game/rules.dart';
import 'package:block_survival/net/protocol.dart';
import 'package:block_survival/world/noise.dart';
import 'package:block_survival/world/palette.dart';
import 'package:block_survival/world/world.dart';
import 'package:flutter_test/flutter_test.dart';

final class _Host implements SimHost {
  _Host(this.world)
    : localAvatar = Avatar(
        id: 'host',
        name: 'Teddy',
        userId: 1,
        spawn: const Vec3(0.5, 1, 0.5),
      );

  @override
  final World world;
  @override
  DayNight dayNight = DayNight();
  @override
  GameRules rules = GameRules.defaults;

  /// the game builds its clock from the rules; the fake does the same on demand
  void useRules(GameRules r) {
    rules = r;
    dayNight = DayNight(r);
  }

  @override
  final Avatar localAvatar;
  final edits = <(int, int, int)>[];
  final broadcasts = <String>[];
  final told = <String>[];
  int dawns = 0;
  int deaths = 0;
  Vec3? respawnedAt;

  @override
  void recordEdit(int x, int y, int z) => edits.add((x, y, z));

  @override
  void broadcastMessage(String text) => broadcasts.add(text);

  @override
  void tell(Avatar a, String text) => told.add('${a.id}: $text');

  @override
  void onDawn(int night) => dawns++;

  @override
  void onLocalDeath() => deaths++;

  @override
  void onLocalRespawn(Vec3 spawn) => respawnedAt = spawn;
}

/// flat grass floor at y = 0 from -r..r
World floor([int r = 40]) {
  final w = World();
  for (var x = -r; x <= r; x++) {
    for (var z = -r; z <= r; z++) {
      w.setBlock(x, 0, z, Block.grass);
    }
  }
  return w;
}

(HostSim, _Host) sim({World? world, int seed = 1}) {
  final host = _Host(world ?? floor());
  return (HostSim(host, rng: Rng(seed)), host);
}

/// a ray from the avatar's eye along +X, dipping so it hits a zombie's body
Ray rayPlusX(Avatar a) => Ray(a.eye(), const Vec3(1, -0.1, 0));

void run(HostSim s, double seconds) {
  final steps = (seconds * 60).round();
  for (var i = 0; i < steps; i++) {
    s.tick(1 / 60);
  }
}

void main() {
  group('night schedule', () {
    test('sunset schedules the groups, they arrive during the night, dawn burns them', () {
      final (s, h) = sim();
      h.dayNight.time = daySeconds - 0.5;
      run(s, 1);
      expect(h.broadcasts, ['Night 1 — they are coming']);
      // 8 zombies on night 1 → round(8 / 4.5) = 2 groups
      expect(s.spawnTimes, hasLength(2));
      run(s, 3);
      expect(s.zombies.liveCount, greaterThan(0));
      expect(s.spawnTimes, hasLength(1));
      h.dayNight.time = daySeconds + nightSeconds - 0.5;
      run(s, 1);
      expect(h.broadcasts.last, 'Dawn — you survived night 1');
      expect(h.dawns, 1);
      expect(s.zombies.liveCount, 0);
      expect(s.spawnTimes, isEmpty);
    });
  });

  test('the admin schedule sets when and how many zombies come', () {
    final (s, h) = sim();
    h.useRules(
      const GameRules(
        daySeconds: 60,
        nightSeconds: 40,
        zombiesFirstNight: 20,
        zombiesPerNight: 0,
        spawnDelaySeconds: 10,
        spawnWindow: 0.5,
      ),
    );
    h.dayNight.time = 59.9;
    run(s, 0.2);
    expect(h.broadcasts, ['Night 1 — they are coming']);
    // 20 zombies → round(20 / 4.5) = 4 groups, the first 10 s after sunset,
    // the rest over half of a 40 s night
    expect(s.spawnTimes, hasLength(4));
    expect(s.spawnTimes.first, closeTo(60 + 10, 0.05));
    expect(s.spawnTimes.last, closeTo(60 + 10 + 15, 0.05));
    run(s, 9);
    expect(s.zombies.liveCount, 0);
    run(s, 2);
    expect(s.zombies.liveCount, greaterThan(0));

    // no zombies at all: nothing is scheduled
    final (s2, h2) = sim();
    h2.useRules(const GameRules(zombiesFirstNight: 0, zombiesPerNight: 0));
    h2.dayNight.time = daySeconds - 0.1;
    run(s2, 1);
    expect(s2.spawnTimes, isEmpty);
    expect(h2.broadcasts, isEmpty);
  });

  group('combat', () {
    test(
      'a sword swing within reach and arc kills a zombie and counts the kill',
      () {
        final (s, h) = sim();
        final me = h.localAvatar;
        me.inventory.add('sword', 1);
        final z = s.zombies.spawn(ZombieKind.basic, 2.5, 1, 0.5);
        s.apply(me, Swing(rayPlusX(me)));
        expect(z.hp, 10);
        expect(z.vx, swordTuning.knockback);
        // behind us: untouched
        final behind = s.zombies.spawn(ZombieKind.basic, -1.5, 1, 0.5);
        s.apply(me, Swing(rayPlusX(me)));
        expect(behind.hp, 30);
        expect(z.state, ZombieState.dead);
        expect(me.kills, 1);
        // no sword in hand: nothing happens behind us either way
        me.slot = 1;
        s.apply(me, Swing(rayPlusX(me)));
        expect(behind.hp, 30);
      },
    );

    test('bare hands and tools still hit, weakly and only up close', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      final near = s.zombies.spawn(ZombieKind.basic, 1.9, 1, 0.5);
      final far = s.zombies.spawn(ZombieKind.basic, 2.9, 1, 0.5);
      s.apply(me, Swing(rayPlusX(me)));
      expect(near.hp, 30 - fistsTuning.damage);
      expect(near.vx, fistsTuning.knockback);
      expect(far.hp, 30); // within a sword's reach, not a fist's
      me.inventory.add('pickaxe_wood', 1);
      s.apply(me, Swing(rayPlusX(me)));
      expect(near.hp, 30 - 2 * fistsTuning.damage);
      // the rifle never swings
      me.inventory.replace([const ItemStack('rifle', 1)]);
      s.apply(me, Swing(rayPlusX(me)));
      expect(near.hp, 30 - 2 * fistsTuning.damage);
    });

    test('the rifle needs a loaded magazine, reloads from ammo and hits the first zombie on the ray', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      me.inventory.add('rifle', 1);
      me.inventory.add('ammo', 40);
      final near = s.zombies.spawn(ZombieKind.soldier, 6.5, 1, 0.5);
      final far = s.zombies.spawn(ZombieKind.basic, 12.5, 1, 0.5);
      // empty magazine: the first trigger pull starts a reload instead
      s.apply(me, Fire(rayPlusX(me)));
      expect(me.magazine, 0);
      expect(me.pendingRounds, rifleMag);
      expect(me.inventory.count('ammo'), 10);
      run(s, RifleTuning.reloadSeconds + 0.1);
      expect(me.magazine, rifleMag);
      s.apply(me, Fire(rayPlusX(me)));
      expect(me.magazine, rifleMag - 1);
      // soldiers take half rifle damage; the one behind is shielded
      expect(near.hp, 80 - RifleTuning.damage * 0.5);
      expect(far.hp, 30);
      // the fire interval gates the next shot
      s.apply(me, Fire(rayPlusX(me)));
      expect(me.magazine, rifleMag - 1);
      run(s, RifleTuning.interval + 0.01);
      s.apply(me, Fire(rayPlusX(me)));
      expect(me.magazine, rifleMag - 2);
    });

    test('a headshot doubles the damage', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      me.inventory.add('rifle', 1);
      me.magazine = rifleMag;
      final z = s.zombies.spawn(ZombieKind.basic, 4.5, 1, 0.5);
      // level from eye height (2.62) the ray passes above the neck (y = 2.5)
      s.apply(me, Fire(Ray(me.eye(), const Vec3(1, 0, 0))));
      expect(z.hp, 30 - RifleTuning.damage * RifleTuning.headshot);
    });

    test('walls stop bullets', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      me.inventory.add('rifle', 1);
      me.magazine = rifleMag;
      for (var y = 1; y <= 3; y++) {
        h.world.setBlock(3, y, 0, Block.stone);
      }
      final z = s.zombies.spawn(ZombieKind.basic, 6.5, 1, 0.5);
      s.apply(me, Fire(rayPlusX(me)));
      expect(me.magazine, rifleMag - 1);
      expect(z.hp, 30);
    });
  });

  group('vitals', () {
    test('an attacking zombie hurts the nearest avatar; toxic ones poison', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      s.zombies.spawn(ZombieKind.toxic, 1.5, 1, 0.5);
      run(s, 1);
      expect(me.health, lessThan(100));
      expect(me.poisonUntil, greaterThan(s.time));
      final afterHit = me.health;
      // the poison keeps ticking after the zombie is gone
      s.zombies.zombies.clear();
      run(s, 1);
      expect(me.health, lessThan(afterHit));
    });

    test('health regenerates after 8 s without damage', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      s.hurt(me, 30, poison: false);
      run(s, 7);
      expect(me.health, 70);
      run(s, 3);
      expect(me.health, greaterThan(70));
      expect(me.health, lessThan(100));
    });

    test('death drops the inventory as a crate and respawns after 5 s at the spawn', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      me.inventory.add('planks', 12);
      me.magazine = 7;
      me.x = 10.5;
      me.z = 3.5;
      s.hurt(me, 500, poison: true);
      s.tick(1 / 60);
      expect(me.dead, isTrue);
      expect(me.deaths, 1);
      expect(h.deaths, 1);
      expect(me.magazine, 0);
      expect(me.poisonUntil, 0);
      expect(me.inventory.count('planks'), 0);
      expect(s.crates.crates, hasLength(1));
      expect(s.crates.crates[0].x, 10.5);
      expect(s.crates.crates[0].y, 1);
      run(s, respawnSeconds + 0.1);
      expect(me.dead, isFalse);
      expect(me.health, AvatarTuning.maxHealth);
      expect(h.respawnedAt, isNotNull);
      expect(me.x, 0.5);
      expect(h.told.last, 'host: Your loot crate is where you fell');
      // walk back and loot it
      me.x = 10.5;
      me.z = 3.5;
      s.apply(
        me,
        Interact(x: 0, y: 0, z: 0, block: 0, crate: s.crates.crates[0].id),
      );
      expect(me.inventory.count('planks'), 12);
      expect(s.crates.crates, isEmpty);
      expect(h.told.last, 'host: Took 12 items');
    });

    test('a remote avatar gets its private state queued instead', () {
      final (s, h) = sim();
      final joiner = Avatar(
        id: 'p1',
        name: 'Sam',
        spawn: const Vec3(20.5, 1, 20.5),
      );
      s.avatars['p1'] = joiner;
      s.hurt(joiner, 500, poison: false);
      s.tick(1 / 60);
      final st = joiner.takeOutbox()!;
      expect(st.health, 0);
      expect(st.dead, isTrue);
      expect(st.respawnIn, respawnSeconds);
      expect(joiner.takeOutbox(), isNull);
      run(s, respawnSeconds + 0.1);
      final back = joiner.takeOutbox()!;
      expect(back.dead, isFalse);
      expect(back.teleport, isNotNull);
      expect(back.teleport!.x, 20.5);
      expect(back.message, 'Back on your feet');
      expect(h.deaths, 0);
    });
  });

  group('blocks', () {
    test('breaking a block within reach drops its item, which is picked up when walked over', () {
      final (s, h) = sim();
      final me = h.localAvatar;
      h.world.setBlock(2, 1, 0, Block.dirt);
      s.apply(me, const BreakBlock(2, 1, 0));
      expect(h.world.getBlock(2, 1, 0), air);
      expect(h.edits, [(2, 1, 0)]);
      expect(s.drops.drops, hasLength(1));
      expect(s.drops.drops[0].item, 'dirt');
      me.x = 2.5;
      run(s, 1.5);
      expect(s.drops.drops, isEmpty);
      expect(me.inventory.count('dirt'), 1);
      // out of reach: ignored
      h.world.setBlock(30, 1, 0, Block.dirt);
      s.apply(me, const BreakBlock(30, 1, 0));
      expect(h.world.getBlock(30, 1, 0), Block.dirt);
    });

    test(
      'placing takes from the slot and refuses occupied or overlapping cells',
      () {
        final (s, h) = sim();
        final me = h.localAvatar;
        me.inventory.add('cobble', 3);
        PlaceBlock at(int x, int y, int z) =>
            PlaceBlock(x: x, y: 0, z: z, nx: 0, ny: y, nz: 0, slot: 0, yaw: 0);
        s.apply(me, at(2, 1, 0));
        expect(h.world.getBlock(2, 1, 0), Block.cobble);
        expect(me.inventory.count('cobble'), 2);
        // same cell again: occupied
        s.apply(me, at(2, 1, 0));
        expect(me.inventory.count('cobble'), 2);
        // where we stand: overlaps the avatar
        s.apply(me, at(0, 1, 0));
        expect(h.world.getBlock(0, 1, 0), air);
        expect(me.inventory.count('cobble'), 2);
      },
    );
  });
}

File: node-server/inertia/sim/__tests__/HostSim.test.ts
import { describe, expect, test } from 'vitest'
import { HostSim, RESPAWN_SECONDS, type Run } from '../HostSim'
import { DEFAULT_RULES } from '../../game/rules'
import { AIR, BLOCK } from '../../world/palette'
import type { SaveData } from '../../game/saveTypes'
import type { ClientMessage } from '../../net/protocol'

const SEED = 4242

function sim(over: Partial<ConstructorParameters<typeof HostSim>[0]> = {}): HostSim {
  return new HostSim({ seed: SEED, rules: DEFAULT_RULES, ...over })
}

/** a block to stand next to: one cell east of the spawn, on the pad's level + 1 */
function beside(s: HostSim): { x: number; y: number; z: number } {
  const sp = s.terrain.spawn()
  return { x: Math.floor(sp.x) + 2, y: Math.floor(sp.y), z: Math.floor(sp.z) }
}

describe('HostSim players', () => {
  test('a new player is welcomed with the seed, the world diff and their spawn', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    const w = s.welcome(a)
    expect(w).toMatchObject({ t: 'welcome', you: 'p1', seed: SEED, rules: { daySeconds: DEFAULT_RULES.daySeconds } })
    expect(w.spawn).toEqual({ x: a.x, y: a.y, z: a.z })
    expect(s.initialState(a)).toMatchObject({ t: 'state', health: 100, spawn: a.spawn })
    expect(s.snapshot().players.map(p => p.name)).toEqual(['Sam'])
  })

  test('private state goes out only when it changed', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    s.initialState(a)
    expect(s.takeState(a)).toBeNull()
    a.inventory.add('dirt', 3)
    expect(s.takeState(a)?.inventory).toBeDefined()
    expect(s.takeState(a)).toBeNull()
  })

  test('a player who leaves keeps their gear for when they come back, at their spawn', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    a.inventory.add('planks', 9)
    a.kills = 4
    a.x += 10
    s.removePlayer('p1')
    expect(s.avatars.size).toBe(0)
    const back = s.addPlayer('p2', 'Sam', 7)
    expect(back.inventory.count('planks')).toBe(9)
    expect(back.kills).toBe(4)
    expect(back.x).toBe(back.spawn.x)
  })
})

describe('HostSim actions', () => {
  test('breaking a block in reach removes it, drops its item and reports the edit', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    const b = beside(s)
    s.world.setBlock(b.x, b.y, b.z, BLOCK.dirt)
    s.takeBlockEdits()
    s.apply(a, { t: 'break', ...b })
    expect(s.world.getBlock(b.x, b.y, b.z)).toBe(AIR)
    expect(s.takeBlockEdits()).toEqual([{ ...b, id: AIR }])
    expect(s.drops.drops.map(d => d.item)).toEqual(['dirt'])
  })

  test('a break out of reach is ignored', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    const b = beside(s)
    s.world.setBlock(b.x + 20, b.y, b.z, BLOCK.dirt)
    s.apply(a, { t: 'break', x: b.x + 20, y: b.y, z: b.z })
    expect(s.world.getBlock(b.x + 20, b.y, b.z)).toBe(BLOCK.dirt)
  })

  test('placing takes one from the slot', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    a.inventory.add('dirt', 2)
    const b = beside(s)
    // the pad's surface under the target cell is solid; place on top of it
    const msg: ClientMessage = { t: 'place', x: b.x, y: b.y - 1, z: b.z, nx: 0, ny: 1, nz: 0, slot: 0, yaw: 0 }
    s.world.setBlock(b.x, b.y - 1, b.z, BLOCK.stone)
    s.world.setBlock(b.x, b.y, b.z, AIR)
    s.apply(a, msg)
    expect(s.world.getBlock(b.x, b.y, b.z)).toBe(BLOCK.dirt)
    expect(a.inventory.count('dirt')).toBe(1)
  })
})

describe('HostSim death and runs', () => {
  test('dying drops the inventory into a crate, records the run, and respawns after the delay', () => {
    const runs: Run[] = []
    const s = sim()
    s.onRun = run => runs.push(run)
    const a = s.addPlayer('p1', 'Sam', 7)
    a.inventory.add('planks', 5)
    a.damage(500)
    s.tick(1 / 30)
    expect(a.dead).toBe(true)
    expect(a.inventory.count('planks')).toBe(0)
    expect(s.crates.crates).toHaveLength(1)
    expect(runs).toEqual([expect.objectContaining({ userId: 7, deaths: 1 })])
    expect(s.takeState(a)).toMatchObject({ dead: true, respawnIn: RESPAWN_SECONDS })
    for (let t = 0; t < RESPAWN_SECONDS + 0.5; t += 1 / 20) s.tick(1 / 20)
    expect(a.dead).toBe(false)
    expect(a.health).toBe(100)
  })

  test('dawn records every player and asks for a save', () => {
    const rules = { ...DEFAULT_RULES, daySeconds: 30, nightSeconds: 30, zombiesFirstNight: 0, zombiesPerNight: 0 }
    const s = sim({ rules })
    const runs: Run[] = []
    let dawns = 0
    s.onRun = run => runs.push(run)
    s.onDawn = () => { dawns++ }
    s.addPlayer('p1', 'Sam', 7)
    s.addPlayer('p2', 'Kim', 8)
    for (let t = 0; t < 61; t += 1 / 20) s.tick(1 / 20)
    expect(dawns).toBe(1)
    expect(runs.map(r => r.userId).sort()).toEqual([7, 8])
    expect(runs[0].nights).toBe(1)
  })
})

describe('HostSim saving', () => {
  test('a save carries the edits and every player, and a new sim resumes from it', () => {
    const s = sim({ ownerId: 7 })
    const owner = s.addPlayer('p1', 'Owner', 7)
    const guest = s.addPlayer('p2', 'Guest', 8)
    owner.inventory.add('log', 3)
    guest.inventory.add('stone', 2)
    owner.x += 3
    const b = beside(s)
    s.world.setBlock(b.x, b.y, b.z, BLOCK.dirt)
    s.apply(owner, { t: 'break', ...b })
    const save: SaveData = JSON.parse(JSON.stringify(s.buildSave()))
    expect(Object.keys(save.players).sort()).toEqual(['7', '8'])

    const again = sim({ restore: save, ownerId: 7 })
    expect(again.world.getBlock(b.x, b.y, b.z)).toBe(AIR)
    const back = again.addPlayer('q1', 'Owner', 7)
    expect(back.inventory.count('log')).toBe(3)
    // the owner resumes where they stood; a visitor would start at their spawn
    expect(back.x).toBeCloseTo(owner.x)
    const visitor = again.addPlayer('q2', 'Guest', 8)
    expect(visitor.inventory.count('stone')).toBe(2)
    expect(visitor.x).toBe(visitor.spawn.x)
  })

  test('the save signature changes when something worth saving happens', () => {
    const s = sim()
    const a = s.addPlayer('p1', 'Sam', 7)
    const before = s.saveSignature()
    a.inventory.add('dirt', 1)
    expect(s.saveSignature()).not.toBe(before)
  })
})
